### ISSUE TITLE
Time limit exceeded with Cloudflare provider

### ISSUE THREAD

**Issue by nunofgs** [janeczku/rancher-letsencrypt#27]
This is the output of `rancher-letsencrypt` with `-debug=1`:

```
7/25/2016 11:04:21 PM[INFO][foo.com, bar.foo.com, biz.foo.com, baz.foo.com, boz.foo.com, qux.foo.com, qix.foo.com] acme: Obtaining bundled SAN certificate
7/25/2016 11:04:22 PM[INFO][foo.com] acme: Could not find solver for: http-01
7/25/2016 11:04:22 PM[INFO][foo.com] acme: Could not find solver for: tls-sni-01
7/25/2016 11:04:22 PM[INFO][foo.com] acme: Trying to solve DNS-01
7/25/2016 11:04:27 PM[INFO][foo.com] Checking DNS record propagation...
7/25/2016 11:04:33 PM[INFO][foo.com] The server validated our request
7/25/2016 11:04:37 PM[INFO][bar.foo.com] acme: Trying to solve DNS-01
7/25/2016 11:04:40 PM[INFO][bar.foo.com] Checking DNS record propagation...
7/25/2016 11:04:47 PM[INFO][bar.foo.com] The server validated our request
7/25/2016 11:04:51 PM[INFO][biz.foo.com] acme: Could not find solver for: tls-sni-01
7/25/2016 11:04:51 PM[INFO][biz.foo.com] acme: Trying to solve DNS-01
7/25/2016 11:04:54 PM[INFO][biz.foo.com] Checking DNS record propagation...
7/25/2016 11:06:59 PM[INFO][baz.foo.com] acme: Could not find solver for: tls-sni-01
7/25/2016 11:06:59 PM[INFO][baz.foo.com] acme: Could not find solver for: http-01
7/25/2016 11:06:59 PM[INFO][baz.foo.com] acme: Trying to solve DNS-01
7/25/2016 11:07:02 PM[INFO][baz.foo.com] Checking DNS record propagation...
7/25/2016 11:09:07 PM[INFO][boz.foo.com] acme: Trying to solve DNS-01
7/25/2016 11:09:09 PM[INFO][boz.foo.com] Checking DNS record propagation...
7/25/2016 11:11:12 PM[INFO][qux.foo.com] acme: Could not find solver for: tls-sni-01
7/25/2016 11:11:12 PM[INFO][qux.foo.com] acme: Trying to solve DNS-01
7/25/2016 11:11:13 PM[INFO][qux.foo.com] Checking DNS record propagation...
7/25/2016 11:13:15 PM[INFO][qix.foo.com] acme: Trying to solve DNS-01
7/25/2016 11:13:17 PM[INFO][qix.foo.com] Checking DNS record propagation...
7/25/2016 11:14:46 PM[INFO][qix.foo.com] The server validated our request
7/25/2016 11:14:48 PMlevel=error msg="[biz.foo.com] Error obtaining certificate: Time limit exceeded. Last error: NS kai.ns.cloudflare.com. did not return the expected TXT record"
7/25/2016 11:14:48 PMlevel=error msg="[baz.foo.com] Error obtaining certificate: Time limit exceeded. Last error: NS kai.ns.cloudflare.com. did not return the expected TXT record"
7/25/2016 11:14:48 PMlevel=error msg="[boz.foo.com] Error obtaining certificate: Time limit exceeded. Last error: NS kai.ns.cloudflare.com. did not return the expected TXT record"
7/25/2016 11:14:48 PMlevel=error msg="[qux.foo.com] Error obtaining certificate: Time limit exceeded. Last error: NS kai.ns.cloudflare.com. did not return the expected TXT record"
```

Any ideas on what's going on here?

**Comment by emilebosch** [comment]
I have the same issue. Not sure what's going on here.

**Comment by nunofgs** [comment]
I'm investigating this a bit and I find that when I add TXT records on my cloudflare account in quick succession, they take longer and longer to show up when I do:

```
$ dig _acme-challenge.bar.foo.com @kai.ns.cloudflare.com TXT
```

So I'm thinking this is Cloudflare limiting me due to requesting a bunch of TXT records one after the other.

@janeczku Any chance you could expose [this timeout](https://github.com/xenolf/lego/blob/master/providers/dns/cloudflare/cloudflare.go#L52) as a flag?

**Comment by nunofgs** [comment]
@emilebosch Just FYI, I was able to generate my certificate successfully. I edited [this timeout](https://github.com/xenolf/lego/blob/master/providers/dns/cloudflare/cloudflare.go#L52) to 1200 (instead of 120) and rebuilt the _rancher-letsencrypt_ image.

**Comment by EugenMayer** [comment]
Started to have issue today while i try to get certificates for 13 sub-domains, before i had 10.
This could explain why people have different results, also considering https://github.com/janeczku/rancher-letsencrypt/issues/12 for example.

Generating certificates takes a really long time actually `Checking DNS record propagation...` takes about 2 minutes per domain. Is this by design? I see TXT records in cloudflare.

This used to be different, it was close to instanct, lets say 10 seconds for all ten, now iam not sure, it looks like 30minutes+. Is that the same issue described here? Iam running with debug=1

**Comment by nunofgs** [comment]
@EugenMayer I believe this may have been a Cloudflare change?

I did notice that the first two subdomains are quick to validate. The third takes 30 seconds or so, then it gets worse after that. It eventually times out at 2 minutes.

Here's how you can debug this:

1. Start this container and wait until it starts to verify the first domain.
2. Go to your cloudflare account and copy the TXT record.
3. Check the TXT record yourself: `dig _acme-challenge.<subdomain>.<domain>.com @<your-name-server>.ns.cloudflare.com TXT`

As you'll find out, it's Cloudflare that's taking a long time to propagate the TXT record for some reason. 😢

**Comment by EugenMayer** [comment]
So basically they probably started to threshold their free service i guess. After waiting for about 30 minutes

```
[INFO][kibana.domain.tld] acme: Could not find solver for: http-01
[INFO][kibana.domain.tld] acme: Could not find solver for: tls-sni-01
[INFO][kibana.domain.tld] acme: Trying to solve DNS-01
[INFO][kibana.domain.tld] Checking DNS record propagation...
[INFO][kibana.domain.tld] The server validated our request
[INFO][auth.domain.tld] acme: Trying to solve DNS-01
[INFO][auth.domain.tld] Checking DNS record propagation...
[INFO][auth.domain.tld] The server validated our request
[INFO][chat.domain.tld] acme: Could not find solver for: tls-sni-01
[INFO][cha.domain.tld] acme: Trying to solve DNS-01
[INFO][chat.domain.tld] Checking DNS record propagation...
[INFO][gems.domain.tld] acme: Trying to solve DNS-01
[INFO][gems.domain.tld] Checking DNS record propagation...
```

after that, no certificate is downloaded /created ```/etc/letsencrypt/production/certs``` is empty.

**Comment by EugenMayer** [comment]
Gave out on this for now, to critical taking you whole infra down because you actually need 3 new certificated, have to re-create the whole stack for that, loosing 10 others and cannot recover. Way to critical (for me) for now

**Comment by emilebosch** [comment]
Well part of the problem is that instead of doing one DNS update it does 2 for each challenge, (add and remove) a better way would be to bundle the DNS update into one batch, and then another to remove the challenges, but unfortunately that doesn't work.

I've been thinking of a different approach which is making a new rancher letsencrypt container: 

That is responsible for:
- requesting
- running a website that responds to the /.wellknown/ challenges
- uploads the certs

Then in rancher loadbalancer, we need to route the * subdomain with  */.wellknown/* to this path to that container

The benefit is that its instant and doesn't require dns, what do you guys think? @EugenMayer @nunofgs

**Comment by EugenMayer** [comment]
lets move this out in a seperate ticket? I guess, while using your great image, i ran into several pitfalls:
 a) adding certificates is real pain ( way to much work )
 b) issues with cloudflare and thus, no free provider offering the DNS-way of authenticating the certs

b)
i do not really care a lot how we solve b), we could just use a high-number port on the host like 64000 and configure a nginx / apache / letsencrypt pseudo server there. When doing the verification, the port shall be given as the verification port and thats basically it. Optionally, we could add a DNS entry for each domain we verify

a) very important issue, there needs to be a way to reconfigure this. For now, i would tend to say that catalogs are the cause of the issue itself and dynamic values shall not be set there. Rather let the user give a path to a yaml file on the host, which configures the certificates to get. Using a rancher-catalog question does imply a lot of bad things, using/writing catalogs a lot, and also using upgrades on them, i would really suppose to do it the other way.

**Comment by emilebosch** [comment]
@EugenMayer would you mind clarifying the following?
The high number port is afaik not possible for verification with lets encrypt https://community.letsencrypt.org/t/support-for-ports-other-than-80-and-443/3419/9

I dont understand what the catalogs have to do with the container? Can you elaborate?

**Comment by EugenMayer** [comment]
@emilebosch one of the main reasons i see people not using your image / catalog is the disability to add certificates when you started the stack. This is also the main reason i stopped using it for now. Re-creating the stack is neither convenient and especially not convinient when issues like timeouts arise during remaking certificates you already had, while you just needed 1 additional.

**Comment by emilebosch** [comment]
@EugenMayer i'm not the owner of this catalog/image. Please check the contributors. 

Also it works perfectly fine, and i think the original author of this container/catalog did a great job. I've used it with much pleasure until cloud flare started acting up with which is out of the control of the original author of this.  

If you needed an extra certificate you just need to upgrade the container and its done, i don;t understand how it could be any simpler than this. The fact that this doesn't work is because cloud flare is rate limiting the DNS propagation requests, which is logical to.

Its a direct consequence of how lets encrypt works now.

**Comment by EugenMayer** [comment]
@emilebosch you are obviously not writing catalogs do you?

Try the following:
1. add letsencrypt stack on 0.2.7 with domain A
2. upgrade the container and add domain B
3. catalog upgrade to 0.3.0

Oh, what happened? Yes, your stack is configured to go for Domain A , no longer for A + B

So "upgrade" is there the really baldy let you fail later on without notice.

**Comment by emilebosch** [comment]
@EugenMayer I'm sorry, i understand that you're frustrated, but feel free to contribute a solution.  As i said i'm not author nor owner of the stack. From the sounds of it. I havent upgraded this catalog yet, but i've been running our staging production with this ssl certs for some time now and we're very happy. For production we just buy wildcard certs.

But as said, please contribute if you have a better solution. It also sounds like it just doesn't keep state between upgrades so that could be a volume issue?

**Comment by janeczku** [comment]
@nunofgs I think we

**Comment by janeczku** [comment]
@nunofgs Having the ability to override the timeout for the propagation check makes sense (see my [comment](https://github.com/xenolf/lego/issues/253#issuecomment-236274225) in the upstream ACME library). But at the end of the day this would merely be a workaround for the shortcomings on  CloudFlare's side. If CloudFlare is either voluntary throttling DNS propagation or having technical issues themselves, then shouldn't you raise this issue with them, @EugenMayer? After all they [claim](https://support.cloudflare.com/hc/en-us/articles/200169036-How-long-does-it-take-for-a-DNS-change-I-made-to-push-out-) that DNS updates will not take longer than the TTL (here 120 sec).
Also, AWS Route 53 is very reasonably priced, so if you are looking for a service more suited for production, it might be worth making the switch to them.

@emilebosch @EugenMayer I think you both suggested to add support for the HTTP based challenge, there is now a ticket for that here https://github.com/janeczku/rancher-letsencrypt/issues/28, and i will look into how this makes the most sense to implement.

**Comment by janeczku** [comment]
Most problems with Cloudflare should be fixed in the new version v0.4.0. Please update your service and reopen this ticket in case the issues persist. Thank you!

**Comment by EugenMayer** [comment]
@janeczku you should really rethink your whole image. with tools like certbot or amce.sh, esp. the latter one, you whole implementation becomes obsolete and they are by all means more powerful, will give you 10 cloud DNS providers out of the box and also are easy to integrate with rancher: ( --reloadcmd )
